pkg/kv/kvtest/store.go

```diff
@@ -0,0 +1,404 @@
+		}
+	})
+
+	t.Run("non_exists", func(t *testing.T) {
+		keyToDel := uniqueKey("missing-key-to-delete")
+		err := store.Delete(ctx, keyToDel)
+		if err != nil {
+			t.Fatalf("delete missing key '%s', err=%v expected nil", keyToDel, err)
+		}
+	})
+}
+
+func testStoreSetIf(t *testing.T, ms makeStore) {
+	ctx := context.Background()
+	store := ms(t, ctx)
+	defer store.Close()
+
+	t.Run("set_first", func(t *testing.T) {
+		key := uniqueKey("set-if-first")
+		val := []byte("v")
+		err := store.SetIf(ctx, key, val, nil)
+		if err != nil {
+			t.Fatalf("SetIf without previous key - key=%s value=%s pred=nil: %s", key, val, err)
+		}
+	})
+
+	t.Run("set_exists", func(t *testing.T) {
+		key := uniqueKey("set-if-exists")
+		val1 := []byte("v1")
+		err := store.Set(ctx, key, val1)
+		if err != nil {
+			t.Fatalf("Set while testing SetIf - key=%s value=%s: %s", key, val1, err)
+		}
+
+		val2 := []byte("v2")
+		err = store.SetIf(ctx, key, val2, val1)
+		if err != nil {
+			t.Fatalf("SetIf with previous value - key=%s value=%s pred=%s: %s", key, val2, val1, err)
+		}
+	})
+
+	t.Run("fail_predicate_value", func(t *testing.T) {
+		key := uniqueKey("fail-predicate-value")
+		val1 := []byte("v1")
+		err := store.Set(ctx, key, val1)
+		if err != nil {
+			t.Fatalf("Set while testing fail predicate - key=%s value=%s: %s", key, val1, err)
+		}
+
+		val2 := []byte("v2")
+		err = store.SetIf(ctx, key, val2, val2)
+		if !errors.Is(err, kv.ErrPredicateFailed) {
+			t.Fatalf("SetIf err=%v - key=%s, value=%s, pred=%s, expected err=%s", err, key, val2, val2, kv.ErrPredicateFailed)
+		}
+	})
+
+	t.Run("fail_predicate_nil", func(t *testing.T) {
+		key := uniqueKey("fail-predicate-nil")
+		val1 := []byte("v1")
+		err := store.Set(ctx, key, val1)
+		if err != nil {
+			t.Fatalf("Set while testing fail predicate - key=%s value=%s: %s", key, val1, err)
+		}
+
+		val2 := []byte("v2")
+		err = store.SetIf(ctx, key, val2, nil)
+		if !errors.Is(err, kv.ErrPredicateFailed) {
+			t.Fatalf("SetIf err=%v - key=%s, value=%s, pred=nil, expected err=%s", err, key, val2, kv.ErrPredicateFailed)
+		}
+	})
+}
+
+func testStoreScan(t *testing.T, ms makeStore) {
+	ctx := context.Background()
+	store := ms(t, ctx)
+	defer store.Close()
+
+	// setup sample data
+	samplePrefix := uniqueKey("scan")
+	const sampleItems = 100
+	sampleData := setupSampleData(t, ctx, store, string(samplePrefix), sampleItems)
+
+	t.Run("full", func(t *testing.T) {
+		scan, err := store.Scan(ctx, samplePrefix)
+		if err != nil {
+			t.Fatal("failed to scan", err)
+		}
+		defer scan.Close()
+
+		var entries []kv.Entry
+		for scan.Next() {
+			ent := scan.Entry()
+			switch {
+			case ent == nil:
+				t.Fatal("scan got nil entry")
+			case ent.Key == nil:
+				t.Fatal("Key is nil while scan item", len(entries))
+			case ent.Value == nil:
+				t.Fatal("Value is nil while scan item", len(entries))
+			}
+			if !bytes.HasPrefix(ent.Key, samplePrefix) {
+				break
+			}
+			entries = append(entries, *ent)
```

Comment:
Can you check that the entries are in-order?

Reply:
Using diff to compare results